flexmeasures/data/models/planning/storage.py

```diff
@@ -368,50 +384,418 @@ def build_device_soc_targets(
+    if soc_maxima is not None:
+        storage_device_constraints["max"] = build_device_soc_values(
+            soc_maxima,
+            soc_at_start,
+            start,
+            end,
+            resolution,
+        )
+
+    storage_device_constraints["max"] = storage_device_constraints["max"].fillna(
+        soc_max_change
+    )
+
+    # limiting max and min to be in the range [soc_min, soc_max]
+    storage_device_constraints["min"] = storage_device_constraints["min"].clip(
+        lower=soc_min_change, upper=soc_max_change
+    )
+    storage_device_constraints["max"] = storage_device_constraints["max"].clip(
+        lower=soc_min_change, upper=soc_max_change
+    )
+
+    return storage_device_constraints
+
+
+def validate_storage_constraints(
+    storage_constraints: pd.DataFrame,
+    soc_at_start: float,
+    min_soc: float,
+    max_soc: float,
+    resolution: timedelta,
+) -> list:
+    """Check that the storage constraints are fulfilled, e.g min <= equals <= max.
+
+    A. Global validation
+        A.1) min <= min_soc
+        A.2) max <= max_soc
+    B. Validation in the same time frame
+        B.1) min <= max
+        B.2) min <= equals
+        B.3) equals <= max
+    C. Validation in different time frames
+        C.1) equals(t) - equals(t-1) <= `derivative max`(t)
+        C.2) `derivative min`(t) <= equals(t) - equals(t-1)
+        C.3) min(t) - max(t-1) <= `derivative max`(t)
+        C.4) max(t) - min(t-1) >= `derivative min`(t)
+        C.5) condition equals(t) - max(t-1) <= `derivative max`(t)
+        C.6) `derivative min`(t) <= equals(t) - min(t-1)
+
+    Args:
+        storage_constraints: pd.DataFrame
+                             dataframe containing the constraints of a storage device
+
+    """
+
+    constraint_violations = []
+
+    ########################
+    # A. Global validation #
+    ########################
+    min_soc = (min_soc - soc_at_start) * timedelta(hours=1) / resolution
+
+    # 1) min <= min_soc
+    mask = ~(storage_constraints["min"] >= min_soc)
+    time_condition_fails = storage_constraints.index[mask]
+
+    for dt in time_condition_fails:
+        value_min = storage_constraints.loc[dt, "min"]
+
+        constraint_violations.append(
+            dict(
+                dt=dt.to_pydatetime(),
+                condition="min <= min_soc",
+                violation=f"min [{value_min}] <= min_soc [{min_soc}]",
+            )
+        )
+
+    # 2) max <= max_soc
+    max_soc = (max_soc - soc_at_start) * timedelta(hours=1) / resolution
+
+    mask = ~(storage_constraints["max"] <= max_soc)
+    time_condition_fails = storage_constraints.index[mask]
+
+    for dt in time_condition_fails:
+        value_max = storage_constraints.loc[dt, "max"]
+
+        constraint_violations.append(
+            dict(
+                dt=dt.to_pydatetime(),
+                condition="max <= max_soc",
+                violation=f"max [{value_max}] <= max_soc [{max_soc}]",
+            )
+        )
```

Comment:
Looks like a great opportunity for refactoring. I made a start for cases A and B. Maybe you can take it over for case C? See https://github.com/FlexMeasures/flexmeasures/pull/702.